File: ShoplyAndroid/swift/Sources/ShoplyCore/Models/UserProfile.swift
//
//  UserProfile.swift
//  ShoplyCore - Android Compatible
//
//  Created by William on 02/11/2025.
//

import Foundation

/// Profil utilisateur - Compatible Android
struct UserProfile: Codable {
    var firstName: String
    var age: Int
    var gender: Gender
    var email: String?
    var createdAt: Date
    var lastWeatherUpdate: Date?
    var preferences: UserPreferences
    
    init(firstName: String = "", age: Int = 0, gender: Gender = .notSpecified, email: String? = nil, createdAt: Date = Date(), preferences: UserPreferences = UserPreferences()) {
        self.firstName = firstName
        self.age = age
        self.gender = gender
        self.email = email
        self.createdAt = createdAt
        self.preferences = preferences
    }
}

public enum Gender: String, Codable, CaseIterable {
    case male = "Homme"
    case female = "Femme"
    case notSpecified = "Non spécifié"
    case other = "Autre"
    
    public var id: String { rawValue }
    
    public init(rawValue: String) {
        switch rawValue {
        case "Homme": self = .male
        case "Femme": self = .female
        case "Non spécifié": self = .notSpecified
        default: self = .other
        }
    }
}

struct UserPreferences: Codable {
    var preferredStyleRawValue: String?
    var favoriteColors: [String] = []
    var comfortLevel: Int = 3 // 1-5
    var styleLevel: Int = 3 // 1-5
    var casualness: Int = 3 // 1-5, 1 = très formel, 5 = très décontracté
    
    var preferredStyle: OutfitType? {
        get {
            guard let rawValue = preferredStyleRawValue else { return nil }
            return OutfitType(rawValue: rawValue)
        }
        set {
            preferredStyleRawValue = newValue?.rawValue
        }
    }
}

File: ShoplyRN/src/models/UserProfile.ts
export enum Gender {
  MALE = 'Homme',
  FEMALE = 'Femme',
  NOT_SPECIFIED = 'Non spécifié',
}

export interface UserPreferences {
  preferredStyleRawValue?: string;
  favoriteColors: string[];
  comfortLevel: number; // 1-5
  styleLevel: number; // 1-5
  casualness: number; // 1-5, 1 = très formel, 5 = très décontracté
}

export interface UserProfile {
  firstName: string;
  dateOfBirth?: Date | string;
  gender: Gender;
  email?: string;
  profilePhotoUri?: string;
  backgroundPhotoUri?: string;
  createdAt: Date | string;
  lastWeatherUpdate?: Date | string;
  preferences: UserPreferences;
}

export const createDefaultUserProfile = (): UserProfile => ({
  firstName: '',
  gender: Gender.NOT_SPECIFIED,
  createdAt: new Date().toISOString(),
  preferences: {
    favoriteColors: [],
    comfortLevel: 3,
    styleLevel: 3,
    casualness: 3,
  },
});

